<a href="https://colab.research.google.com/github/Evgeniy-luka/main/blob/main/proect_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Анализ игровой индустрии с 2000 по 2013 год

- Автор: Ляпустин Е.
- Дата: начало 22.09.2025

### Цели и задачи проекта

Изучить развитие игровой индустрии за период с 2000 по 2013 год включительно

### Описание данных

<font color='#777778'>Здесь приведите описание данных.</font>

Описание данных
Данные /datasets/new_games.csv содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:

Name — название игры.

Platform — название платформы.

Year of Release — год выпуска игры.

Genre — жанр игры.

NA sales — продажи в Северной Америке (в миллионах проданных копий).

EU sales — продажи в Европе (в миллионах проданных копий).

JP sales — продажи в Японии (в миллионах проданных копий).

Other sales — продажи в других странах (в миллионах проданных копий).

Critic Score — оценка критиков (от 0 до 100).

User Score — оценка пользователей (от 0 до 10).

Rating — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

### Содержимое проекта

Загрузка и знакомство с данными

Проверка ошибок в данных и их предобработка:

- Названия, или метки, столбцов датафрейма

- Типы данных

- Наличие пропусков в данных

- Явные и неявные дубликаты в данных

Фильтрация данных период с 2000 по 2013 год включительно

Категоризация данных:

- по оценкам пользователей

- по оценкам критиков

Выделение топ-7 платформ по количеству игр, выпущенных за весь требуемый период

## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv('/datasets/new_games.csv')

- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [ ]:
start_row = df.shape[0] # Фиксирую колличество строк в исходном Датасете
start_row

16956

In [ ]:
df.head()

Name Platform  Year of Release         Genre  NA sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  EU sales JP sales  Other sales  Critic Score User Score Rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

- Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.
- Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

1. Объем данных

Количество строк 16956

Количество столбцов 11

2. Данные соответствуют описанию

3. Пропуски в данных

Имеются пропуски в данных необходимо более глубокое изучение

В столбце Critic Score более 50 % пропусков, Rating и User Score по 40 % пропусков, Year of Release 1,62% пропусков,
Name и Genre по 0.11% соответственно, в остальных столбцах пропуски отсутствуют.     

4. Типы данных

Соответствие типов данных: float64(4), object(7)

Столбцы Name, Platform, Genre, NA sales, Other sales, Critic Score, Rating -  имеют верный тип данных.

Столбцы EU sales, JP sales, User Score  текущий - тип object, требуемый тип float64 для проведения с этими столбцами вычислений.

Столбец Year of Release - имеет тип данных float64, можно его изменить для красоты на тип int

5. Другие особенности и замечания

Названия столбцов необходимо привести к стилю snake case, а названия должны быть в нижнем регистре


---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [ ]:
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [ ]:
# приведем названием колонок в правильный формат
df.columns = df.columns.str.lower().str.replace(" ", "_")

In [ ]:
df.head()

name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  eu_sales jp_sales  other_sales  critic_score user_score rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16956 non-null  object 
 6   jp_sales         16956 non-null  object 
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       10152 non-null  object 
 10  rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

In [ ]:
df[['eu_sales', 'jp_sales','na_sales', 'user_score']] = df[['eu_sales', 'jp_sales','na_sales', 'user_score']].apply(pd.to_numeric, errors='coerce')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16950 non-null  float64
 6   jp_sales         16952 non-null  float64
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float64
 10  rating           10085 non-null  object 
dtypes: float64(7), object(4)
memory usage: 1.4+ MB


###  Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [ ]:
df.isna().sum().sort_values(ascending=False) # Считаю и сортирую колличество пропусков

user_score         9268
critic_score       8714
rating             6871
year_of_release     275
eu_sales              6
jp_sales              4
name                  2
genre                 2
platform              0
na_sales              0
other_sales           0
dtype: int64

In [ ]:
# Процент пропуск в колонках
df.isna().mean().sort_values(ascending=False) * 100

user_score         54.659118
critic_score       51.391838
rating             40.522529
year_of_release     1.621845
eu_sales            0.035386
jp_sales            0.023590
name                0.011795
genre               0.011795
platform            0.000000
na_sales            0.000000
other_sales         0.000000
dtype: float64

- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.

Имеются пропуски в следующих столбцах:

- critic_score и user_score имеют более чем 50% отсутствует, считаю что изменять и удалять ничего не надо. Пропуски заполню Nan чтобы в дальнейшем так же информацию об отсутствии значений выделить в отдельную категорию.

- rating имеют более чем 40% пропуски в данных, что так же говорит о удаление или замена только искозят данные и иследование будет недостоверны. Пропуски переименую в Undefined чтобы в дальнейшем так же информацию об отсутствии значений выделить в отдельную категорию.

- year_of_release имеют незначительные пропуски менее 2%. Приемлемая потеря для чистоты данных.

- name и genre имеют по 2 пропуск. Поскольку пропусков очень мало, скорее всего, это случайные ошибки. Лучше всего эти строки удалить.

<div class="alert alert-info">
Возможные причины возникновения пропусков столбцах user_score и critic_score:

- игры старые и в то время оценки ставить просто негде было интернет не был так распространен;

- игры вышли недавно и не получили оценки;

- игры были не популярны и не были массовыми;

- технические сбои при сборе данных.

Возможные причины возникновения пропусков столбце rating:

- система рейтинга была введена в 1994 году и игры до этого момента могли не получить рейтинг;

- не всем играм пристваивают рейтинг это на усмотрение платформы которая его выпустила;

- технические или человеческие ошибки при внесении или заполнении базы.

Возможные причины возникновения пропусков столбце year_of_release:

- игра могла выйти в разных странах в разное время;

- технические или человеческие ошибки при внесении или заполнении базы.

Возможная причина возникновения пропусков столбцах не превышающих 0,3 процента могли возникнуть в результате технической или человеческой при заполнении базы или ее выгрузке

In [ ]:
df.head(20)

name platform  year_of_release  \
0                                     Wii Sports      Wii           2006.0   
1                              Super Mario Bros.      NES           1985.0   
2                                 Mario Kart Wii      Wii           2008.0   
3                              Wii Sports Resort      Wii           2009.0   
4                       Pokemon Red/Pokemon Blue       GB           1996.0   
5                                         Tetris       GB           1989.0   
6                          New Super Mario Bros.       DS           2006.0   
7                                       Wii Play      Wii           2006.0   
8                      New Super Mario Bros. Wii      Wii           2009.0   
9                                      Duck Hunt      NES           1984.0   
10                                    Nintendogs       DS           2005.0   
11                                 Mario Kart DS       DS           2005.0   
12                   Pokemon Gold/Pokemon Silver       GB           1999.0   
13                                       Wii Fit      Wii           2007.0   
14                            Kinect Adventures!     X360           2010.0   
15                                  Wii Fit Plus      Wii           2009.0   
16                            Grand Theft Auto V      PS3           2013.0   
17                 Grand Theft Auto: San Andreas      PS2           2004.0   
18                             Super Mario World     SNES           1990.0   
19  Brain Age: Train Your Brain in Minutes a Day       DS           2005.0   

           genre  na_sales  eu_sales  jp_sales  other_sales  critic_score  \
0         Sports     41.36     28.96      3.77         8.45          76.0   
1       Platform     29.08      3.58      6.81         0.77           NaN   
2         Racing     15.68     12.76      3.79         3.29          82.0   
3         Sports     15.61     10.93      3.28         2.95          80.0   
4   Role-Playing     11.27      8.89     10.22         1.00           NaN   
5         Puzzle     23.20      2.26      4.22         0.58           NaN   
6       Platform     11.28      9.14      6.50         2.88          89.0   
7           Misc     13.96      9.18      2.93         2.84          58.0   
8       Platform     14.44      6.94      4.70         2.24          87.0   
9        Shooter     26.93      0.63      0.28         0.47           NaN   
10    Simulation      9.05     10.95      1.93         2.74           NaN   
11        Racing      9.71      7.47      4.13         1.90          91.0   
12  Role-Playing      9.00      6.18      7.20         0.71           NaN   
13        Sports      8.92      8.03      3.60         2.15          80.0   
14          Misc     15.00      4.89      0.24         1.69          61.0   
15        Sports      9.01      8.49      2.53         1.77          80.0   
16        Action      7.02      9.09      0.98         3.96          97.0   
17        Action      9.43      0.40      0.41        10.57          95.0   
18      Platform     12.78      3.75      3.54         0.55           NaN   
19          Misc      4.74      9.20      4.16         2.04          77.0   

    user_score rating  
0          8.0      E  
1          NaN    NaN  
2          8.3      E  
3          8.0      E  
4          NaN    NaN  
5          NaN    NaN  
6          8.5      E  
7          6.6      E  
8          8.4      E  
9          NaN    NaN  
10         NaN    NaN  
11         8.6      E  
12         NaN    NaN  
13         7.7      E  
14         6.3      E  
15         7.4      E  
16         8.2      M  
17         9.0      M  
18         NaN    NaN  
19         7.9      E

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16950 non-null  float64
 6   jp_sales         16952 non-null  float64
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float64
 10  rating           10085 non-null  object 
dtypes: float64(7), object(4)
memory usage: 1.4+ MB


In [ ]:
df[df['genre'].isna()] # Считаю что их необходимо удалить

name platform  year_of_release genre  na_sales  eu_sales  jp_sales  \
661    NaN      GEN           1993.0   NaN      1.78      0.53      0.00   
14439  NaN      GEN           1993.0   NaN      0.00      0.00      0.03   

       other_sales  critic_score  user_score rating  
661           0.08           NaN         NaN    NaN  
14439         0.00           NaN         NaN    NaN

In [ ]:
df = df.dropna(subset=['genre']) # удалю строки с nan в столбце genre т.к. они не представлют большого интереса

In [ ]:
df = df.dropna(subset=['year_of_release']) # удалю строки с nan
df['year_of_release'] = df['year_of_release'].astype('int64') # преобразую в int

In [ ]:
# заменяю на среднее значение в зависимости от названия платформы и года выхода игры
df['jp_sales'] = df['jp_sales'].fillna(df.groupby(['platform', 'year_of_release'])['jp_sales'].transform('mean'))

In [ ]:
df['eu_sales'] = df['eu_sales'].fillna(df.groupby(['platform', 'year_of_release'])['eu_sales'].transform('mean'))

In [ ]:
df['na_sales'] = df['na_sales'].fillna(df.groupby(['platform', 'year_of_release'])['na_sales'].transform('mean'))

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16679 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16679 non-null  object 
 1   platform         16679 non-null  object 
 2   year_of_release  16679 non-null  int64  
 3   genre            16679 non-null  object 
 4   na_sales         16679 non-null  float64
 5   eu_sales         16679 non-null  float64
 6   jp_sales         16679 non-null  float64
 7   other_sales      16679 non-null  float64
 8   critic_score     8085 non-null   float64
 9   user_score       7558 non-null   float64
 10  rating           9901 non-null   object 
dtypes: float64(6), int64(1), object(4)
memory usage: 1.5+ MB


- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

In [ ]:
unique_count = df['genre'].unique()
unique_count

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION', 'SHOOTER',
       'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE', 'SIMULATION',
       'PUZZLE', 'STRATEGY'], dtype=object)

имеются неявные дубликаты в столбце genre, будем приводить к нижнему регистру

In [ ]:
df['genre'] = df['genre'].str.lower()

In [ ]:
unique_count = df['platform'].unique()
unique_count

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

в столбце platform нет явных или неявных дубликатов

In [ ]:
unique_count = df['rating'].unique()
unique_count

array(['E', nan, 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

в столбце rating имеется неявный дубликат K-A и E - это одинаковые рейтинги. K-A (Kids to Adults) - старый рейтинг, использовался до 1998 года, а E (Everyone) - современный эквивалент, введен в 1998 году. Приведу ка я их к единому рейтингу Е.

In [ ]:
df['rating'] = df['rating'].replace('K-A', 'E')

In [ ]:
unique_count = df['rating'].unique()
unique_count

array(['E', nan, 'M', 'T', 'E10+', 'AO', 'EC', 'RP'], dtype=object)

In [ ]:
unique_count = df['year_of_release'].unique()
unique_count

array([2006, 1985, 2008, 2009, 1996, 1989, 1984, 2005, 1999, 2007, 2010,
       2013, 2004, 1990, 1988, 2002, 2001, 2011, 1998, 2015, 2012, 2014,
       1992, 1997, 1993, 1994, 1982, 2016, 2003, 1986, 2000, 1995, 1991,
       1981, 1987, 1980, 1983])

в столбце year_of_release нет явных или неявных дубликатов

- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

In [ ]:
nunique_count = df[['name', 'platform', 'year_of_release', 'genre']].nunique()
nunique_count

name               11426
platform              31
year_of_release       37
genre                 12
dtype: int64

In [ ]:
df.duplicated().sum() # Перед удалением перепроверяю сколько у нас полных дубликатов

235

Для определения явных дубликатов выделил 4 столбца для их поиска 'name', 'platform', 'year_of_release', 'genre'.
Явные дубликаты считаю необходимым удалить для очищения данных.

In [ ]:
df = df.drop_duplicates(subset=['name', 'platform', 'year_of_release', 'genre'])

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16443 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16443 non-null  object 
 1   platform         16443 non-null  object 
 2   year_of_release  16443 non-null  int64  
 3   genre            16443 non-null  object 
 4   na_sales         16443 non-null  float64
 5   eu_sales         16443 non-null  float64
 6   jp_sales         16443 non-null  float64
 7   other_sales      16443 non-null  float64
 8   critic_score     7982 non-null   float64
 9   user_score       7462 non-null   float64
 10  rating           9767 non-null   object 
dtypes: float64(6), int64(1), object(4)
memory usage: 1.5+ MB


- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

In [ ]:
final_rows  = df.shape[0] # Фиксирую колличество строк после очишения Датасета
final_rows

16443

In [ ]:
del_rows = start_row - final_rows # Кол-во удаленных строк
del_rows

513

In [ ]:
doly_del_rows = del_rows / start_row * 100
doly_del_rows

3.0254777070063694

В процессе предобработки данных были выявлены и устранены неявные и явные дубликаты, а также пропущенные значения. Все некорректные записи были либо удалены, либо приведены к единому среднему.

Исходный датасет содержал 16956 записей. После очистки было удалено 513 строк, что составляет 3% от общего объема данных. Такой  процент потерь позволяет считать полученный датасет репрезентативным и готов для дальнейшего анализа.


---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [ ]:
df_actual = df[(df['year_of_release'] >= 2000) & (df['year_of_release'] <= 2013)]

In [ ]:
df_actual['year_of_release'].min(), df_actual['year_of_release'].max()

(2000, 2013)

In [ ]:
df_actual.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12780 entries, 0 to 16954
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             12780 non-null  object 
 1   platform         12780 non-null  object 
 2   year_of_release  12780 non-null  int64  
 3   genre            12780 non-null  object 
 4   na_sales         12780 non-null  float64
 5   eu_sales         12780 non-null  float64
 6   jp_sales         12780 non-null  float64
 7   other_sales      12780 non-null  float64
 8   critic_score     7168 non-null   float64
 9   user_score       6482 non-null   float64
 10  rating           8722 non-null   object 
dtypes: float64(6), int64(1), object(4)
memory usage: 1.2+ MB


---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [ ]:
def user_score_group(row):
    x = row['user_score']
    if x < 3:
        return 'low_score'
    elif x >= 3 and x < 8:
        return 'avg_score'
    elif x >= 8 and x <= 10:
        return 'high_score'
    else:
        return 'unknown'

In [ ]:
df_actual = df_actual.copy()
df_actual['user_score_category'] = df_actual.apply(user_score_group, axis=1)

In [ ]:
df_actual.head()

name platform  year_of_release     genre  na_sales  \
0             Wii Sports      Wii             2006    sports     41.36   
2         Mario Kart Wii      Wii             2008    racing     15.68   
3      Wii Sports Resort      Wii             2009    sports     15.61   
6  New Super Mario Bros.       DS             2006  platform     11.28   
7               Wii Play      Wii             2006      misc     13.96   

   eu_sales  jp_sales  other_sales  critic_score  user_score rating  \
0     28.96      3.77         8.45          76.0         8.0      E   
2     12.76      3.79         3.29          82.0         8.3      E   
3     10.93      3.28         2.95          80.0         8.0      E   
6      9.14      6.50         2.88          89.0         8.5      E   
7      9.18      2.93         2.84          58.0         6.6      E   

  user_score_category  
0          high_score  
2          high_score  
3          high_score  
6          high_score  
7           avg_score

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [ ]:
def critic_score_group(row):
    x = row['critic_score']
    if x < 30:
        return 'low_score'
    elif x >= 30 and x < 80:
        return 'avg_score'
    elif x >= 80 and x <= 100:
        return 'high_score'
    else:
        return 'unknown'

In [ ]:
df_actual['critic_score_category'] = df_actual.apply(critic_score_group, axis=1)

In [ ]:
df_actual.head(28)

name platform  year_of_release  \
0                                     Wii Sports      Wii             2006   
2                                 Mario Kart Wii      Wii             2008   
3                              Wii Sports Resort      Wii             2009   
6                          New Super Mario Bros.       DS             2006   
7                                       Wii Play      Wii             2006   
8                      New Super Mario Bros. Wii      Wii             2009   
10                                    Nintendogs       DS             2005   
11                                 Mario Kart DS       DS             2005   
13                                       Wii Fit      Wii             2007   
14                            Kinect Adventures!     X360             2010   
15                                  Wii Fit Plus      Wii             2009   
16                            Grand Theft Auto V      PS3             2013   
17                 Grand Theft Auto: San Andreas      PS2             2004   
19  Brain Age: Train Your Brain in Minutes a Day       DS             2005   
20                 Pokemon Diamond/Pokemon Pearl       DS             2006   
23                            Grand Theft Auto V     X360             2013   
24                   Grand Theft Auto: Vice City      PS2             2002   
25                 Pokemon Ruby/Pokemon Sapphire      GBA             2002   
26   Brain Age 2: More Training in Minutes a Day       DS             2005   
27                   Pokemon Black/Pokemon White       DS             2010   
28                        Gran Turismo 3: A-Spec      PS2             2001   
29                Call of Duty: Modern Warfare 3     X360             2011   
32                       Call of Duty: Black Ops     X360             2010   
33                           Pokemon X/Pokemon Y      3DS             2013   
34                    Call of Duty: Black Ops II      PS3             2012   
35                    Call of Duty: Black Ops II     X360             2012   
36                Call of Duty: Modern Warfare 2     X360             2009   
37                Call of Duty: Modern Warfare 3      PS3             2011   

           genre  na_sales  eu_sales  jp_sales  other_sales  critic_score  \
0         sports     41.36     28.96      3.77         8.45          76.0   
2         racing     15.68     12.76      3.79         3.29          82.0   
3         sports     15.61     10.93      3.28         2.95          80.0   
6       platform     11.28      9.14      6.50         2.88          89.0   
7           misc     13.96      9.18      2.93         2.84          58.0   
8       platform     14.44      6.94      4.70         2.24          87.0   
10    simulation      9.05     10.95      1.93         2.74           NaN   
11        racing      9.71      7.47      4.13         1.90          91.0   
13        sports      8.92      8.03      3.60         2.15          80.0   
14          misc     15.00      4.89      0.24         1.69          61.0   
15        sports      9.01      8.49      2.53         1.77          80.0   
16        action      7.02      9.09      0.98         3.96          97.0   
17        action      9.43      0.40      0.41        10.57          95.0   
19          misc      4.74      9.20      4.16         2.04          77.0   
20  role-playing      6.38      4.46      6.04         1.36           NaN   
23        action      9.66      5.14      0.06         1.41          97.0   
24        action      8.41      5.49      0.47         1.78          95.0   
25  role-playing      6.06      3.90      5.38         0.50           NaN   
26        puzzle      3.43      5.35      5.32         1.18          77.0   
27  role-playing      5.51      3.17      5.65         0.80           NaN   
28        racing      6.85      5.09      1.87         1.16          95.0   
29       shooter      9.04      4.24      0.13         1.32          88.0   
32       s

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [ ]:
user_score_group = df_actual.groupby('user_score_category')['name'].count()
user_score_group

user_score_category
avg_score     4080
high_score    2286
low_score      116
unknown       6298
Name: name, dtype: int64

In [ ]:
critic_score_group = df_actual.groupby('critic_score_category')['name'].count()
critic_score_group

critic_score_category
avg_score     5422
high_score    1691
low_score       55
unknown       5612
Name: name, dtype: int64

- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [ ]:
top_platforms = df_actual['platform'].value_counts().head(7)
top_platforms

PS2     2127
DS      2120
Wii     1275
PSP     1180
X360    1121
PS3     1086
GBA      811
Name: platform, dtype: int64

---

## 5. Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

Для исследования игровой индустрии за период 2000-2013 годов была проведена проверка ошибок в данных и их предобработка, состоящая из следующие этапов:

1. Названия, или метки, столбцов датафрейма
Исходный датасет содержал 16,956 записей с информацией о играх, платформах, продажах и оценках

Выявлены пропуски в ключевых столбцах: critic_score (51%), user_score (40%), rating (40%)

2. Корректировка структуры данных
Приведены названия столбцов к стилю snake_case

Выполнено преобразование типов данных: годы выпуска к целочисленному типу, оценки к числовому формату

Обработаны текстовые значения в числовые поля

3. Наличие пропусков в данных

Удалены строки с пропусками значений в годах выпуска игр 1,62 %

Удалены незначительные пропуски в названиях игр и жанрах 0.01%

Пропуски в данных с количеством проданных копий игры в том или ином регионе были заполнены средним значение в зависимости от названия платформы и года выхода игры

Для оценок критиков и пользователей созданы отдельные категории для пропущенных значений "unknown"

4. Явные и неявные дубликаты в данных
Выявлены и исправлены неявные дубликаты в жанрах из-за разного регистра написания

Удалены явные дубликаты сравнимые по столбцам 'name', 'platform', 'year_of_release', 'genre'

Объединены исторические варианты рейтингов ESRB (K-A → E)

Проверены уникальные значения платформ и годов выпуска на наличие дубликатов

5. Фильтрация и категоризация данных
Создан целевой датафрейм df_actual с играми за период 2000-2013 годов (12780  строк)

Реализована категоризация оценок пользователей и критиков по трем уровням

Выделен топ-7 платформ по количеству выпущенных игр.